### DISCUSSION TITLE
calculateCountsFromAbundance with StringTie Data

### DISCUSSION THREAD

**Discussion by erichpeterson** [kvittingseerup/IsoformSwitchAnalyzeR#97] [Q&A] [answered]
Hi,
I am a bit confused about the _calculateCountsFromAbundance_ parameter when importing StringTie data to be fixed and then re-exported again (to be imported by DESeq2. 

Here are the experiments I have performed on the same dataset:
1. I have created a count matrix using the StringTie prepDE.py script, imported it into DESeq2 using the instructions provided by that group (i.e., _DESeqDataSetFromMatrix_), and performed builtin DESeq2 normalization on the data.
2. I have used _tximport_, to import the data into DESeq2 using the instructions from the DESeq2 vignette (in which it claims that the bias correction will be done automatically when using the _DESeqDataSetFromTximport_), and performed builtin DESeq2 normalization on the data.
3. Finally, I have imported the data into _IsoformSwithAnalyzeR_ using the default parameters. I then used _importRdata_ to correct the StringTie problems with annotations, then called _extractGeneExpression_ to get the count matrix. Lastly, I imported this matrix into DESeq2 using the _DESeqDataSetFromMatrix_ method and normalized the data.

The resulting counts for methods (1) and (2) were very close, whereas, the results for (3) were much reduced compared to the others. I tried setting the _calculateCountsFromAbundance_ to FALSE and re-ran, and this time the counts were very similar to (1) and (2). 

So, I am a bit confused, given that the documentation for IsoformSwitchAnalyzeR recommends keeping the value set to the default (TRUE). Am I doing something wrong?

Thanks

**Comment by kvittingseerup** [comment, marked as answer]
Hi 

Thanks for reaching out.

Could you elaborate on what you mean by "much reduced compared to the others"?

In the mean time I can elaborate a bit on what I can already say. Scaling abundance from transcripts widely recognised as the better approach - including by the authors of DESeq2 as can be seen from this section of the [tximport vignette ](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#Import_transcript-level_estimates) (tximport is created by same authors as DESeq2) as well as their [benchmarking article](https://f1000research.com/articles/7-952). That is why it is default in IsoformSwitchAnalyzeR.

Specifically we need to do something "that corrects for changes to the average transcript length across samples" (a quote from [the tximport manual](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#downstream-dge-in-bioconductor)). This can be accomplished in two ways:

1. By rescaling scaling the actual counts (as IsoformSwitchAnalyzeR does)
2. By adding the extra factor as an offset in the DE model which is what `DESeqDataSetFromTximport()` does.

So the reason why you see a difference between 2 & 3 is that the actual values are not re-scaled in 2 - they are added as an offset in the model.

Lastly part of the reason why there are differences is that when you do the scaling you have a couple of options (please refere to the tximport for all of them). Here IsoformSwitchAnalyzeR uses "ScaledTPM" because it is suitable both for differential expression and diffrential usage analysis (see figure 1 of [this](https://f1000research.com/articles/7-952)).

I hope this makes sense.

Cheers
Kristoffer

**Reply by erichpeterson** [reply]
Hi,
Thanks so much for the timely reply.

What I mean by "much reduced" is that for (1) and (2) the count after importation and DESeq2 normalization are within a few counts of each other, whereas with (3) the counts are maybe 1/3 or 1/4 of (1) and (2). 

So, I think I might understand a bit more now. Correct me if I am wrong, but basically, the difference in counts between (1) and (2), and (3) is because the default StringTie _tximport_ function is doing the **counts + length** method, whereas, by default, the _importIsoformExpression_ is doing the **scaledTPM** method.

Thanks again,
Erich

**Reply by kvittingseerup** [reply]
Thanks for elaborating. On that note you can also double check that the total number of counts are not changing - meaning they are just redistributed aka scaled.

And yes. If you do a transcript level DE analysis on 1, 2, and 3 you will probably see that 2 and 3 are very similar whereas 1 is different as it have no scaling done.

**Reply by erichpeterson** [reply]
Sorry, for beating a dead horse, but could you confirm that executing the following IsoformSwitchAnalyzeR code:

```r
StringTieQuant <- importIsoformExpression(
	parentDir = "./"
)

myDesign <- read.table(design_file, header=TRUE)
myDesign

aSwitchList <- importRdata(
	isoformCountMatrix   = StringTieQuant$counts,
	isoformRepExpression = StringTieQuant$abundance,
	designMatrix         = myDesign,
	isoformExonAnnoation = "./output_merged.gtf",
	fixStringTieAnnotationProblem = TRUE,
	addAnnotatedORFs = FALSE,
	estimateDifferentialGeneRange = FALSE
)

#export genes
df_gene <- extractGeneExpression(aSwitchList, addGeneNames = FALSE)
write.table(df_gene, "gene_quant.csv", row.names = FALSE, quote = FALSE)
```

Then importing into DESEq2 like below is OK to do, given that in the DESeQ2 tutorial says the following: "The DESeq2 model internally corrects for library size, so transformed or normalized values such as counts scaled by library size should not be used as input."

```r
countData <- as.matrix(read.csv("gene_quant.csv", row.names="gene_id", header = TRUE, sep = ' ', as.is = TRUE, check.names = FALSE))
# colData/condition specified here
dds <- DESeqDataSetFromMatrix(countData = countData,
                                 colData = colData,
                                 design = ~ condition)
```

Thanks

**Reply by kvittingseerup** [reply]
That will be perfectly fine :-). The counts obtained from `extractGeneExpression()` are neither transformed, normalized, or scaled by library size.

With the code above you will however have a problem as the file you write (with `write.table()`) and the one you read in again (with `read.csv()`) are not the same. Additionally you naturally you don't need to write the counts to a file but can just use `df_gene` directly as input to `DESeqDataSetFromMatrix()`.
